src/Microsoft.AspNet.Mvc.Razor/Precompilation/PrecompilationTagHelperTypeResolver.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNet.Razor.Runtime.TagHelpers;
+using Microsoft.CodeAnalysis;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc.Razor.Precompilation
+{
+    /// <summary>
+    /// <see cref="TagHelperDescriptorResolver"/> used during Razor precompilation.
+    /// </summary>
+    public class PrecompilationTagHelperTypeResolver : TagHelperTypeResolver
+    {
+        private readonly object _assemblyLookupLock = new object();
+        private readonly Dictionary<string, IEnumerable<ITypeInfo>> _assemblyLookup
+            = new Dictionary<string, IEnumerable<ITypeInfo>>(StringComparer.OrdinalIgnoreCase);
+        private readonly CodeAnalysis.Compilation _compilation;
+        private INamedTypeSymbol _openGenericDictionaryTypeSymbol;
+
+        public PrecompilationTagHelperTypeResolver([NotNull] CodeAnalysis.Compilation compilation)
+        {
+            _compilation = compilation;
+        }
+
+        /// <inheritdoc />
+        protected override IEnumerable<ITypeInfo> GetExportedTypes([NotNull] AssemblyName assemblyName)
+        {
+            lock (_assemblyLookup)
+            {
+                IEnumerable<ITypeInfo> result;
+                if (!_assemblyLookup.TryGetValue(assemblyName.Name, out result))
```

Comment:
It's once per precompilation. Although we don't necessarily do any work until we determine a view actually needs to be recompiled (i.e. view compilation gets cached the same runtime compilation does).